transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,471 @@
+        connection_id: ConnectionHandle,
+        event: EndpointEvent,
+    },
+    /// Instruct the endpoint to transmit a packet on its UDP socket.
+    Transmit(Transmit),
+}
+
+#[derive(Debug)]
+pub struct TransportChannel<C: Crypto> {
+    tx: mpsc::UnboundedSender<ToEndpoint<C>>,
+    rx: mpsc::Receiver<Result<QuicMuxer<C>, QuicError>>,
+    port: u16,
+    ty: SocketType,
+}
+
+impl<C: Crypto> TransportChannel<C> {
+    pub fn dial(
+        &mut self,
+        addr: SocketAddr,
+        public_key: C::PublicKey,
+    ) -> oneshot::Receiver<Result<QuicMuxer<C>, QuicError>> {
+        let (tx, rx) = oneshot::channel();
+        let msg = ToEndpoint::Dial {
+            addr,
+            public_key,
+            tx,
+        };
+        self.tx.unbounded_send(msg).expect("endpoint has crashed");
+        rx
+    }
+
+    pub fn poll_incoming(
+        &mut self,
+        cx: &mut Context,
+    ) -> Poll<Option<Result<QuicMuxer<C>, QuicError>>> {
+        Pin::new(&mut self.rx).poll_next(cx)
+    }
+
+    pub fn port(&self) -> u16 {
+        self.port
+    }
+
+    pub fn ty(&self) -> SocketType {
+        self.ty
+    }
+}
+
+#[derive(Debug)]
+pub struct ConnectionChannel<C: Crypto> {
+    id: ConnectionHandle,
+    tx: mpsc::UnboundedSender<ToEndpoint<C>>,
+    rx: mpsc::Receiver<ConnectionEvent>,
+    port: u16,
+    max_datagrams: usize,
+}
+
+impl<C: Crypto> ConnectionChannel<C> {
+    pub fn poll_channel_events(&mut self, cx: &mut Context) -> Poll<ConnectionEvent> {
+        match Pin::new(&mut self.rx).poll_next(cx) {
+            Poll::Ready(Some(event)) => Poll::Ready(event),
+            Poll::Ready(None) => panic!("endpoint has crashed"),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+
+    pub fn send_endpoint_event(&mut self, event: EndpointEvent) {
+        let msg = ToEndpoint::ConnectionEvent {
+            connection_id: self.id,
+            event,
+        };
+        self.tx.unbounded_send(msg).expect("endpoint has crashed")
+    }
+
+    pub fn send_transmit(&mut self, transmit: Transmit) {
+        let msg = ToEndpoint::Transmit(transmit);
+        self.tx.unbounded_send(msg).expect("endpoint has crashed")
+    }
+
+    pub fn port(&self) -> u16 {
+        self.port
+    }
+
+    pub fn max_datagrams(&self) -> usize {
+        self.max_datagrams
+    }
+}
+
+#[derive(Debug)]
+struct EndpointChannel<C: Crypto> {
+    rx: mpsc::UnboundedReceiver<ToEndpoint<C>>,
+    tx: mpsc::Sender<Result<QuicMuxer<C>, QuicError>>,
+    port: u16,
+    max_datagrams: usize,
+    connection_tx: mpsc::UnboundedSender<ToEndpoint<C>>,
+}
+
+impl<C: Crypto> EndpointChannel<C> {
+    pub fn poll_next_event(&mut self, cx: &mut Context) -> Poll<Option<ToEndpoint<C>>> {
+        Pin::new(&mut self.rx).poll_next(cx)
+    }
+
+    pub fn create_connection(
+        &self,
+        id: ConnectionHandle,
+    ) -> (ConnectionChannel<C>, mpsc::Sender<ConnectionEvent>) {
+        let (tx, rx) = mpsc::channel(12);
```

Comment:
What is the reasoning for the constant `12`?

Reply:
empirical testing. I printlned when it was blocked and 12 seemed to work well for the test